pkg/executor/sortexec/sort.go

```diff
@@ -17,86 +17,92 @@ package sortexec
 import (
 	"container/heap"
 	"context"
-	"errors"
+	"time"
 
+	"github.com/pingcap/errors"
 	"github.com/pingcap/failpoint"
 	"github.com/pingcap/tidb/pkg/executor/internal/exec"
 	"github.com/pingcap/tidb/pkg/expression"
 	"github.com/pingcap/tidb/pkg/planner/util"
 	"github.com/pingcap/tidb/pkg/sessionctx/variable"
+	"github.com/pingcap/tidb/pkg/types"
 	"github.com/pingcap/tidb/pkg/util/chunk"
 	"github.com/pingcap/tidb/pkg/util/disk"
 	"github.com/pingcap/tidb/pkg/util/memory"
-	"github.com/pingcap/tidb/pkg/util/sqlkiller"
 )
 
 // SortExec represents sorting executor.
 type SortExec struct {
 	exec.BaseExecutor
 
 	ByItems    []*util.ByItems
-	Idx        int
 	fetched    bool
 	ExecSchema *expression.Schema
 
 	// keyColumns is the column index of the by items.
 	keyColumns []int
 	// keyCmpFuncs is used to compare each ByItem.
 	keyCmpFuncs []chunk.CompareFunc
-	// rowChunks is the chunks to store row values.
-	rowChunks *chunk.SortedRowContainer
+
+	curPartition *sortPartition
+
+	// We can't spill if size of data is lower than the limit
+	spillLimit int64
 
 	memTracker  *memory.Tracker
 	diskTracker *disk.Tracker
 
-	// PartitionList is the chunks to store row values for partitions. Every partition is a sorted list.
-	PartitionList []*chunk.SortedRowContainer
+	// sortPartitions is the chunks to store row values for partitions. Every partition is a sorted list.
+	sortPartitions []*sortPartition
+	cursors        []*dataCursor
 
 	// multiWayMerge uses multi-way merge for spill disk.
 	// The multi-way merge algorithm can refer to https://en.wikipedia.org/wiki/K-way_merge_algorithm
 	multiWayMerge *multiWayMerge
 	// spillAction save the Action for spill disk.
-	spillAction *chunk.SortAndSpillDiskAction
+	spillAction *sortPartitionSpillDiskAction
+
+	helper                *spillHelper
+	enableTmpStorageOnOOM bool
```

Comment:
> Why add this?

`variable.EnableTmpStorageOnOOM.Load()` controls if the spill can be triggered, we need to check this variable each time we call `switchToNewSortPartition` function, it's better to save this variable in advance so that we can avoid to repeatedly load that variable each time.

Reply:
repeatedly load that variable is ok..  there are no many partitions